### PR TITLE
feat: Add custom http status code and headers support for cloud funct…

### LINKED ISSUE #9479: Return HTTP response status codes from Cloud Code
### New Feature / Enhancement Checklist

- Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
- Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
- Before posting search [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Current Limitation
The HTTP status codes returned by cloud code in the response's header can not be determined in the cloud functions.

### Feature / Enhancement Description
Ability to set the HTTP status code in the response header of a cloud function call.

### Example Use Case
Set a 401 status code in the response header when an unauthenticated user tries to access unauthorised data.

### Alternatives / Workarounds
Currently we can manually set any information in the response body and mimic the status code functionality by manually parsing and handling the data and logic.

### 3rd Party References
All backends should have this ability.

### PR DESCRIPTION
**Pull request by hajjarjoseph**
## Pull Request

  - Report security issues [confidentially](https://github.com/parse-community/parse-server/security/policy).
  - Any contribution is under this [license](https://github.com/parse-community/parse-server/blob/alpha/LICENSE).
  - Link this pull request to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

  ## Issue

  Closes: #9479

  ## Approach

  Modified `FunctionsRouter.createResponseObject` to detect `__httpResponse: true` in cloud function return values and pass through `status` and `headers` to the existing PromiseRouter infrastructure. Fully backwards compatible.
  
  edited new approach:
  
Added CloudResponse class and pass it as second argument to cloud functions (req, res). When res.status() or res.set() are called, FunctionsRouter.createResponseObject passes status and headers to the existing PromiseRouter infrastructure. The res argument is optional - existing functions using only (req) continue to work unchanged.

  ## Tasks

  - [x] Add tests
  - [x] Add changes to documentation (guides, repository pages, code comments)
  - [x] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
  - [ ] Add new Parse Error codes to Parse JS SDK

## Summary by CodeRabbit

* **New Features**
  * Cloud Functions can optionally customize HTTP responses (set status codes and headers via an Express-like response object).

* **Documentation**
  * README updated with usage examples, chaining behavior, and security considerations for response headers; note that the response argument is optional.

* **Tests**
  * Extensive new tests covering status/header setting, validation, combined responses, error paths, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by parseplatformorg** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/acinader/pr-checks/63c4e5fd-d589-4dfe-a28c-10cdb116daa5?refs=71fba5fd440c5195016fd3e24332194dc3b76f4e&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

**Comment by mtrezza** [comment]
I'm not a fan of this implementation; it looks like manipulation a server-internal property. Exposing a response object alongside the request object would be a cleaner approach.

**Comment by hajjarjoseph** [comment]
> I'm not a fan of this implementation; it looks like manipulation a server-internal property. Exposing a response object alongside the request object would be a cleaner approach.



> I'm not a fan of this implementation; it looks like manipulation a server-internal property. Exposing a response object alongside the request object would be a cleaner approach.

Changed my approach:

Added CloudResponse class and pass it as second argument to cloud functions (req, res). When res.status() or res.set() are called, FunctionsRouter.createResponseObject passes status and headers to the existing PromiseRouter infrastructure. The res argument is optional - existing functions using only (req) continue to work unchanged.

  Example:
  ```
Parse.Cloud.define('myFunction', (req, res) => {
    res.status(201).set('X-Custom-Header', 'value');
    return { message: 'Created' };
  });
```

I do not want the bounty for this feature implementation, would just be happy if I get a tiny contribution to a project I have been using regularly!

File: src/Routers/FunctionsRouter.js [hunk 1]
```diff
@@ -9,6 +9,41 @@ import { jobStatusHandler } from '../StatusHandler';
...
+  }
+
+  getStatus() {
+    return this._status;
+  }
+
+  getHeaders() {
+    return this._headers;
+  }
+}
```
Comment (hajjarjoseph) [inline src/Routers/FunctionsRouter.js:63]: - Added validation: status code range (100-599), header name must be string, header value must be defined
  - Added defensive copy of headers object

Documentation:
  - Added "Cloud Functions HTTP Response" section to README with usage examples, API reference, and security considerations

Tests:
  - Added edge case tests for: multiple status() calls, multiple set() calls, invalid status codes, invalid header names/values

File: README.md [hunk 2]
```diff
@@ -771,6 +771,55 @@ Logs are also viewable in Parse Dashboard.
...
+The `res` argument is optional. Existing cloud functions using only `(req) => {}` continue to work unchanged.
+
+### Security Considerations
+
+The `set()` method allows setting arbitrary HTTP headers. Be cautious when setting security-sensitive headers such as:
+- CORS headers (`Access-Control-Allow-Origin`, `Access-Control-Allow-Credentials`)
+- `Set-Cookie`
+- `Location` (redirects)
+- Authentication headers (`WWW-Authenticate`)
+
```
[the remark this answers is not in the document]
Comment (hajjarjoseph) [reply README.md:823]: Done

File: src/Routers/FunctionsRouter.js [hunk 3]
```diff
@@ -9,6 +9,47 @@ import { jobStatusHandler } from '../StatusHandler';
...
```
[the remark this answers is not in the document]
Comment (hajjarjoseph) [reply src/Routers/FunctionsRouter.js:63]: Done

**Comment by mtrezza** [comment]
See https://github.com/parse-community/parse-server/pull/9980

**Comment by hajjarjoseph** [comment]
> ## [Codecov](https://app.codecov.io/gh/parse-community/parse-server/pull/9979?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community) Report
> 
> ✅ All modified and coverable lines are covered by tests. ✅ Project coverage is 92.55%. Comparing base ([`d5e76b0`](https://app.codecov.io/gh/parse-community/parse-server/commit/d5e76b01db2b4eeb22a0bb5a04347a89209aa822?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community)) to head ([`71fba5f`](https://app.codecov.io/gh/parse-community/parse-server/commit/71fba5fd440c5195016fd3e24332194dc3b76f4e?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community)). ⚠️ Report is 1 commits behind head on alpha.
> Additional details and impacted files
> 
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/parse-community/parse-server/pull/9979?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=parse-community).
> 🚀 New features to boost your workflow:

@mtrezza did you check the changes here? Will you go with #9980? Thank you :-)

**Comment by mtrezza** [comment]
@hajjarjoseph I think we'll go with https://github.com/parse-community/parse-server/pull/9980, it also contains a new express-style syntax for composing the response. Would you want to test out that PR before we merge - if you have a use case for it?

**Comment by hajjarjoseph** [comment]
> @hajjarjoseph I think we'll go with #9980, it also contains a new express-style syntax for composing the response. Would you want to test out that PR before we merge - if you have a use case for it?
 
All good 👍
